vendors/st/boards/stm32l475_discovery/ports/common_io/iot_i2c.c

```diff
@@ -0,0 +1,744 @@
+    HAL_GPIO_Init( DISCOVERY_I2Cx_SCL_SDA_GPIO_PORT, &xGpio_init_structure );
+
+    /* Configure the I2C peripheral */
+    DISCOVERY_I2Cx_CLK_ENABLE();
+
+    DISCOVERY_I2Cx_FORCE_RESET();
+
+    DISCOVERY_I2Cx_RELEASE_RESET();
+
+    /* Enable and set I2Cx Interrupt to a lower priority */
+    HAL_NVIC_SetPriority( DISCOVERY_I2Cx_EV_IRQn, 0x0F, 0 );
+    HAL_NVIC_EnableIRQ( DISCOVERY_I2Cx_EV_IRQn );
+}
+
+void HAL_I2C_MspDeInit( I2C_HandleTypeDef * hi2c )
+{
+    GPIO_InitTypeDef xGpio_init_structure;
+
+    /* Configure I2C Tx, Rx as alternate function */
+    xGpio_init_structure.Pin = DISCOVERY_I2Cx_SCL_PIN | DISCOVERY_I2Cx_SDA_PIN;
+    HAL_GPIO_DeInit( DISCOVERY_I2Cx_SCL_SDA_GPIO_PORT, xGpio_init_structure.Pin );
+
+    DISCOVERY_I2Cx_SCL_SDA_GPIO_CLK_DISABLE();
+
+    DISCOVERY_I2Cx_CLK_DISABLE();
+}
+
+void HAL_I2C_MasterTxCpltCallback( I2C_HandleTypeDef * hi2c )
+{
+    IotI2CDescriptor_t * pDescriptor = _toDescriptor( hi2c );
+    BaseType_t higherPriorityTaskWoken = pdFALSE;
+
+    /* Subtract not transferred yet. */
+    pDescriptor->usTransmittedTxBytes -= hi2c->XferCount;
+
+    if( pDescriptor->xCallback != NULL )
+    {
+        pDescriptor->xCallback( eI2CCompleted, pDescriptor->pvUserContext );
+    }
+
+    xSemaphoreGiveFromISR( pDescriptor->xSemphr, &higherPriorityTaskWoken );
+    portYIELD_FROM_ISR( higherPriorityTaskWoken );
+}
+
+void HAL_I2C_MasterRxCpltCallback( I2C_HandleTypeDef * hi2c )
+{
+    IotI2CDescriptor_t * pDescriptor = _toDescriptor( hi2c );
+    BaseType_t higherPriorityTaskWoken = pdFALSE;
+
+    /* Subtract not transferred yet. */
+    pDescriptor->usReceivedRxBytes -= hi2c->XferCount;
+
+    if( pDescriptor->xCallback != NULL )
+    {
+        pDescriptor->xCallback( eI2CCompleted, pDescriptor->pvUserContext );
+    }
+
+    xSemaphoreGiveFromISR( pDescriptor->xSemphr, &higherPriorityTaskWoken );
+    portYIELD_FROM_ISR( higherPriorityTaskWoken );
+}
+
+void HAL_I2C_ErrorCallback( I2C_HandleTypeDef * hi2c )
+{
+    IotI2CDescriptor_t * pDescriptor = _toDescriptor( hi2c );
+    BaseType_t higherPriorityTaskWoken = pdFALSE;
+
+    if( pDescriptor->xCallback != NULL )
+    {
+        pDescriptor->xCallback( eI2CCompleted, pDescriptor->pvUserContext );
+    }
+
+    xSemaphoreGiveFromISR( pDescriptor->xSemphr, &higherPriorityTaskWoken );
+    portYIELD_FROM_ISR( higherPriorityTaskWoken );
+}
+
+/*---------------------Private functions---------------------*/
+
+static int32_t _toI2cStatus( HAL_StatusTypeDef halStatus,
+                             uint8_t operation )
+{
+    switch( halStatus )
+    {
+        case HAL_OK:
+            return IOT_I2C_SUCCESS;
+
+        case HAL_ERROR:
+            return operation == _I2C_READ_OP ? IOT_I2C_READ_FAILED : IOT_I2C_WRITE_FAILED;
+
+        case HAL_BUSY:
+            return IOT_I2C_BUSY;
+
+        case HAL_TIMEOUT:
+            return IOT_I2C_BUS_TIMEOUT;
+    }
+
+    /* TODO: it is better to assert(0) here. */
```

Comment:
Why?

Reply:
I was thinking since _toI2cStatus is a private function, all the cases of "halStatus" parameter should be covered by the switch block above. So this line of code should be not reachable.

Reply:
The code should be refactored then.

Reply:
Any suggestion? Assert looks the best to me.